core/src/error/suggest.rs

```diff
@@ -0,0 +1,49 @@
+//! Suggest existing symbols that are similar to what the user typed in case of various "symbol not
+//! found" errors, such as [super::EvalError::UnboundIdentifier] or
+//! [super::EvalError::FieldMissing].
+//!
+//! The current implementation uses the normalized Damereau-Levenshtein edit distance to find the
+//! closest match.
+
+use strsim::normalized_damerau_levenshtein;
+
+/// The minimum similarity between the user's input and an existing symbol for the symbol to be
+/// considered a serious candidate. The current threshold is rather low, because short words with
+/// edit distance 1 such as `bar` and `bare` might have a similarity which isn't that high.
+pub const MIN_SIMILARITY: f64 = 0.60;
+
+/// Find the closest match to the user input in the given list of symbols, to be used as a
+/// suggestion. If there is an exact match (modulo casing), it is returned immediately. Otherwise,
+/// the symbol with the highest normalized Damerau-Levenshtein similarity is returned, as long as
+/// the similarity is above the fixed threshold [MIN_SIMILARITY].
+pub fn find_best_match<'syms, 'input, S, I>(
+    symbols: &'syms [S],
+    user_input: &'input I,
+) -> Option<&'syms str>
+where
+    S: AsRef<str>,
+    I: AsRef<str>,
+{
+    let mut max = std::f64::MIN;
+    let mut arg_max: Option<&'syms str> = None;
+
+    if symbols.is_empty() {
+        return None;
+    }
+
+    for sym in symbols {
+        // If there is an exact match (modulo case), return it immediately.
+        if sym.as_ref().to_lowercase() == user_input.as_ref().to_lowercase() {
+            return Some(sym.as_ref());
+        }
+
+        let similarity = normalized_damerau_levenshtein(sym.as_ref(), user_input.as_ref());
```

Comment:
(nit): Since we're first comparing things in a case-insensitive manner, would it make sense to also normalise the case when computing the distance?

Reply:
I just copied rustc here, and [they do keep the original case] for edit distance computations. I imagine the idea is, if there is an exact match modulo casing, you might have messed up the casing (like a camelCase error) with high probability, but if there's more than casing, then casing is considered as any other edit/mispelling (for example we don't consider keystrokes that are closer to have somehow a smaller cost in the edit distance - of course that would be hard to do across all keyboard layouts, and that's maybe why).

Rustc doing it doesn't mean it's the best approach necessarily, but I guess it at least means it's not entirely stupid :man_shrugging: